rama-http/src/service/web/router.rs

```diff
@@ -0,0 +1,512 @@
+        self.match_route(path, matcher, service)
+    }
+
+    /// register a nested router under a prefix.
+    ///
+    /// The prefix is used to match the request path and strip it from the request URI.
+    pub fn sub<I, T>(self, prefix: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let path = format!("{}/{}", prefix.trim_end_matches(['/']), "{*nest}");
+        let matcher = HttpMatcher::custom(true);
+
+        let nested_router_service = NestedRouterService {
+            prefix: prefix.to_owned(),
+            nested: service.into_endpoint_service().boxed(),
+        };
+
+        self.match_route(&path, matcher, nested_router_service)
+    }
+
+    /// add a route to the router with it's matcher and service.
+    pub fn match_route<I, T>(
+        mut self,
+        path: &str,
+        matcher: HttpMatcher<State, Body>,
+        service: I,
+    ) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let service = service.into_endpoint_service().boxed();
+
+        if let Ok(matched) = self.routes.at_mut(path) {
+            matched.value.push((matcher, service));
+        } else {
+            self.routes
+                .insert(path, vec![(matcher, service)])
+                .expect("Failed to add route");
+        }
+
+        self
+    }
+
+    /// use the provided service when no route matches the request.
+    pub fn not_found<I, T>(mut self, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        self.not_found = Some(service.into_endpoint_service().boxed());
+        self
+    }
+}
+
+struct NestedRouterService<State> {
+    prefix: String,
+    nested: BoxService<State, Request, Response, Infallible>,
+}
+
+impl<State: std::fmt::Debug> std::fmt::Debug for NestedRouterService<State> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("NestedRouterService")
+            .field("prefix", &self.prefix)
+            .finish()
+    }
+}
+
+struct ParentUriParams(UriParams);
+
+impl Clone for ParentUriParams {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+impl ParentUriParams {
+    fn merge_with_child(&self, child_params: &UriParams) -> UriParams {
+        let mut merged = self.0.clone();
+        for (key, value) in child_params.iter() {
+            merged.insert(key.to_owned(), value.to_owned());
+        }
+        merged
+    }
+}
+
+impl<State> Service<State, Request> for NestedRouterService<State>
+where
+    State: Clone + Send + Sync + 'static,
+{
+    type Response = Response;
+    type Error = Infallible;
+
+    async fn serve(
+        &self,
+        mut ctx: Context<State>,
+        mut req: Request,
+    ) -> Result<Self::Response, Self::Error> {
+        if let Some(param) = ctx.get::<UriParams>() {
+            let nested_params = param.clone();
+
+            if let Some(nest) = nested_params.get("nest") {
```

Comment:
it should be an internal server error in case it is not available as you are the one defining this param, so how could it ever not be in there?